src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/AvTrace.cs

```diff
@@ -336,32 +334,28 @@ public string Trace( TraceEventType type, int eventId, string message, string[]
                 type,
                 eventId,
                 traceMessage,
-                arrayList.ToArray() );
+                combinedList.ToArray()); //Cannot avoid the alloc here, no ROS<object> overload
```

Comment:
You could change the `List<object> combinedList` to an `object[]` with length `parameters.Length * 2`; then you won't need to create a list, which creates an underlying array (`_items`) and creates an array in `.ToArray()`.

Reply:
It would need to be `Math.Min(labels.Length - 1, parameters.Length) * 2` (unless we can assume those are always equal to the same value).

Reply:
As @bgrainger mentions, that is why I didn't do this change and put comment instead. 

Theoretically I could use `Max(...) * 2` instead and just allocate one bigger array, it would work out for the better. `TraceEvent` uses `string.Format` under the hood, which doesn't care how many params in the array, as long as it is not less.

Reply:
Done, I've also delayed the alloc, according to my tests of all possible trace variations, all the trace outputs work as expected, and are same.

https://github.com/dotnet/runtime/blob/99c9f5b95a68a5e864436f9ec0726eed759350bd/src/libraries/System.Diagnostics.TraceSource/src/System/Diagnostics/TraceListener.cs#L374-L377 ref for null arrays.